astrix-remoting/src/main/java/com/avanza/astrix/remoting/server/AstrixServiceActivatorImpl.java

```diff
@@ -97,46 +113,58 @@ public ServiceInvocationMonitors(DynamicBooleanProperty serviceMonitorEnabled, S
 	
 	private static class PublishedServiceMethod<T> {
 		private final ServiceInvocationMonitors serviceInvocationMonitors;
+		private final List<InvocationExecutionWatcher> invocationExecutionWatchers;
 		private final Method serviceMethod;
 		private final AstrixObjectSerializer objectSerializer;
 		private final T service;
-		
-		public PublishedServiceMethod(ServiceInvocationMonitors serviceInvocationMonitors, Method method, AstrixObjectSerializer objectSerializer, T service) {
+		private final AstrixTraceProvider astrixTraceProvider;
+
+		public PublishedServiceMethod(
+				ServiceInvocationMonitors serviceInvocationMonitors,
+				List<InvocationExecutionWatcher> invocationExecutionWatchers,
+				Method method,
+				AstrixObjectSerializer objectSerializer,
+				T service,
+				AstrixTraceProvider astrixTraceProvider
+		) {
 			this.serviceInvocationMonitors = serviceInvocationMonitors;
+			this.invocationExecutionWatchers = invocationExecutionWatchers;
 			this.serviceMethod = method;
 			this.objectSerializer = objectSerializer;
 			this.service = service;
+			this.astrixTraceProvider = astrixTraceProvider;
 		}
 		
 		private AstrixServiceInvocationResponse timeInvocation(AstrixServiceInvocationRequest request, int version) {
 			return serviceInvocationMonitors.monitorServiceInvocation(() -> invoke(request, version)).call();
 		}
 		
 		private AstrixServiceInvocationResponse invoke(AstrixServiceInvocationRequest request, int version) {
+			Runnable afterInvocationWatchers = InvocationExecutionWatcher.apply(invocationExecutionWatchers, request.getHeaders());
 			try {
 				return invokeService(request, version);
 			} catch (Exception e) {
 				Throwable exceptionThrownByService = resolveException(e);
 				AstrixServiceInvocationResponse invocationResponse = new AstrixServiceInvocationResponse();
 				invocationResponse.setExceptionMsg(exceptionThrownByService.getMessage());
-				invocationResponse.setCorrelationId(UUID.randomUUID().toString());
+				invocationResponse.setCorrelationId(getCorrelationId(astrixTraceProvider, request));
 				if (exceptionThrownByService instanceof ServiceInvocationException) {
 					invocationResponse.setException(this.objectSerializer.serialize(exceptionThrownByService, version));
 				} else {
 					invocationResponse.setThrownExceptionType(exceptionThrownByService.getClass().getName());
 				}
 				logger.info(String.format("Service invocation ended with exception. request=%s correlationId=%s", request, invocationResponse.getCorrelationId()), exceptionThrownByService);
 				return invocationResponse;
+			} finally {
+				afterInvocationWatchers.run();
```

Comment:
Do we need to consider that afterInvocationWatchers.run() might throw an exception?